src/stream/src/executor/join/hash_join.rs

```diff
@@ -667,19 +668,43 @@ impl JoinEntryState {
         key: PkType,
         value: StateValueType,
     ) -> Result<&mut StateValueType, JoinEntryError> {
+        let mut removed = false;
+        if !enable_strict_consistency() {
+            // strict consistency is off, let's remove existing (if any) first
+            if let Some(old_value) = self.cached.remove(&key) {
+                self.kv_heap_size.sub(&key, &old_value);
+                removed = true;
+            }
+        }
```

Comment:
Why do we need to `remove` first? Can't we simply match the `Result` of `try_insert`?

Reply:
Oh, I missed prior discussion

Reply:
Wait, I still don't get it. Is it because of `kv_heap_size`? What if we `sub` it on error?

Reply:
Insert will take the ownership of the value, and if failed, we have no way to get the value back. If to try insert first, we have to clone the value.

Reply:
Oh, this is typical! But where do we need the new value after insertion? :eyes:

Reply:
Oh you remind me. Maybe it can be better to just ignore the new value when there's one existing.

Reply:
As discussed offline, let's keep the current impl.